File: swift/Sources/TokenTrackerCore/Scanners/KimiScanner.swift
//
//  KimiScanner.swift
//  TokenTrackerCore
//
//  移植自 scanners/kimi.py：~/.kimi-code/server/events/session_*.jsonl
//  （事件日志）+ kimi-cli ~/.kimi/sessions/（通用 JSONL 兜底）。
//  turn.step.completed 的 usage 是每步增量 → 直接累加，不重复计数。
//

import Foundation

public struct KimiScanner: ScannerAdapter {
    public let name = "kimi"
    public let detail = "~/.kimi-code/server/events/session_*.jsonl"
    public let journalDir: String
    public let cliDir: String

    public init(journalDir: String, cliDir: String) {
        self.journalDir = expandPath(journalDir)
        self.cliDir = expandPath(cliDir)
    }

    private func isDirectory(_ path: String) -> Bool {
        var isDir: ObjCBool = false
        return FileManager.default.fileExists(atPath: path, isDirectory: &isDir) && isDir.boolValue
    }

    public func detect() -> Bool {
        isDirectory(journalDir) || isDirectory(cliDir)
    }

    private func parseTS(_ raw: Any?) -> Int64 {
        if let n = raw as? NSNumber, CFGetTypeID(n) != CFBooleanGetTypeID() {
            let d = n.doubleValue
            return Int64(d < 1e12 ? d * 1000 : d)
        }
        if let s = raw as? String { return parseISODateMs(s) ?? 0 }
        return 0
    }

    private func jsonlFiles(under base: String, matching predicate: (String) -> Bool) -> [String] {
        var out: [String] = []
        if let enumerator = FileManager.default.enumerator(atPath: base) {
            for case let entry as String in enumerator where predicate(entry) {
                out.append((base as NSString).appendingPathComponent(entry))
            }
        }
        return out.sorted()
    }

    private func scanJournal(_ store: UsageStore, _ prices: PriceTable,
                             cursor: inout [String: Any], full: Bool) throws -> ScanOutcome {
        var outcome = ScanOutcome()
        guard isDirectory(journalDir) else { return outcome }
        let files = jsonlFiles(under: journalDir) {
            ($0 as NSString).lastPathComponent.hasPrefix("session_")
                && $0.hasSuffix(".jsonl")
        }
        for path in files {
            if !full && !fingerprintChanged(cursor: cursor, path: path) { continue }
            guard let statKey = StatKey(path: path) else { continue }
            outcome.files += 1
            let filename = (path as NSString).lastPathComponent
            let sessionID = String(filename.dropFirst("session_".count).dropLast(".jsonl".count))
            var project = ""
            var modelHint = ""
            var title: String?
            for (_, obj) in iterJSONL(path) {
                let kind = obj["kind"] as? String
                let envelope = obj["envelope"] as? [String: Any] ?? [:]
                let payload = envelope["payload"] as? [String: Any] ?? [:]
                let eventType = envelope["type"] as? String ?? ""
                let activityTS = parseTS(jsonOrAny(envelope["timestamp"], obj["time"]))
                if kind == "event" && eventType == "tool.call.started" {
                    let call = payload["toolCall"] as? [String: Any] ?? payload
                    let rawName = jsonOrString(call["name"], call["toolName"], call["tool"])
                    let callID = jsonOrString(call["id"], call["toolCallId"], payload["toolCallId"])
                    if !rawName.isEmpty {
                        let change = try store.recordActivity(
                            agent: name, srcKey: "\(sessionID)|tool|\(callID.isEmpty ? String(describing: obj["seq"] ?? "") : callID)",
                            rawName: rawName, sessionID: sessionID,
                            turnID: payload["turnId"] as? String ?? "", callID: callID,
                            startedAt: activityTS == 0 ? nil : activityTS,
                            sourceKind: "kimi_journal",
                            arguments: jsonOrAny(call["args"], call["arguments"], call["input"]))
                        outcome.activityAdded += change.added
                        outcome.activityUpdated += change.updated
                    }
                } else if kind == "event" && eventType == "tool.result" {
                    let callID = jsonOrString(payload["toolCallId"], payload["callId"], payload["id"])
                    var status = ActivityNormalizer.status(jsonOrAny(payload["status"], payload["error"]))
                    if status == "unknown" { status = payload["error"] == nil ? "success" : "error" }
                    outcome.activityUpdated += try store.completeActivity(
                        agent: name, callID: callID, status: status,
                        endedAt: activityTS == 0 ? nil : activityTS,
                        durationMs: (payload["durationMs"] as? NSNumber)?.int64Value)
                }
                if title == nil && kind == "event" && envelope["type"] as? String == "turn.started" {
                    if let prompt = payload["prompt"] as? String,
                       !prompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                        title = cleanSessionTitle(prompt)
                    }
                }
                if kind == "event" && envelope["type"] as? String == "event.session.created" {
                    let session = payload["session"] as? [String: Any] ?? [:]
                    let meta = session["metadata"] as? [String: Any] ?? [:]
                    if let cwd = meta["cwd"] as? String, !cwd.isEmpty { project = cwd }
                    continue
                }
                guard kind == "event",
                      envelope["type"] as? String == "turn.step.completed",
                      let usage = payload["usage"] as? [String: Any] else { continue }
                let inp = jsonInt(usage["inputOther"])      // 非缓存输入增量
                let outp = jsonInt(usage["output"])
                let cr = jsonInt(usage["inputCacheRead"])
                let cw = jsonInt(usage["inputCacheCreation"])
                if inp + outp + cr + cw == 0 { continue }
                var m = payload["model"]
                if let dict = m as? [String: Any] { m = dict["id"] }
                if let mStr = m as? String, !mStr.isEmpty { modelHint = mStr }
                // 事件日志不携带模型字段，默认 kimi-code（k3 家族）
                let model = modelHint.isEmpty ? "kimi-code" : modelHint
                let ts = parseTS(jsonOrAny(envelope["timestamp"], obj["time"]))
                let key = "\(sessionID)|step|\((obj["seq"] as? NSNumber).map { $0.stringValue } ?? "None")"
                let cost = prices.cost(for: model, input: inp, output: outp,
                                       cacheRead: cr, cacheWrite: cw)
                outcome.added += try store.putEvent(
                    tool: name, srcKey: key, sessionID: sessionID, project: project,
                    ts: ts, model: model, input: inp, output: outp,
                    cacheRead: cr, cacheWrite: cw, cost: cost)
            }
            cursor[path] = statKey.asDict
            if let title {
                try store.setSessionTitle(tool: name, sessionID: sessionID, title: title)
            }
        }
        return outcome
    }

    private func scanCLI(_ store: UsageStore, _ prices: PriceTable,
                         cursor: inout [String: Any], full: Bool) throws -> ScanOutcome {
        var outcome = ScanOutcome()
        guard isDirectory(cliDir) else { return outcome }
        let files = jsonlFiles(under: cliDir) { $0.hasSuffix(".jsonl") }
        for path in files {
            if !full && !fingerprintChanged(cursor: cursor, path: path) { continue }
            guard let statKey = StatKey(path: path) else { continue }
            outcome.files += 1
            let filename = (path as NSString).lastPathComponent
            for (lineno, obj) in iterJSONL(path) {
                guard let usage = obj["usage"] as? [String: Any] else { continue }
                let inp = jsonOrInt(usage["input"], usage["input_tokens"])
                let outp = jsonOrInt(usage["output"], usage["output_tokens"])
                if inp + outp == 0 { continue }
                let model = obj["model"] as? String ?? ""
                let cost = prices.cost(for: model, input: inp, output: outp)
                outcome.added += try store.putEvent(
                    tool: name, srcKey: "cli|\(path)|\(lineno)",
                    sessionID: String(filename.dropLast(".jsonl".count)),
                    project: (path as NSString).deletingLastPathComponent,
                    ts: parseTS(obj["timestamp"]), model: model,
                    input: inp, output: outp, cost: cost)
            }
            cursor[path] = statKey.asDict
        }
        return outcome
    }

    public func scan(_ store: UsageStore, _ prices: PriceTable, full: Bool) throws -> ScanOutcome {
        var cursor = try store.getScanCursor(tool: name)
        let effectiveFull = full || activityNeedsBackfill(cursor)
        let journal = try scanJournal(store, prices, cursor: &cursor, full: effectiveFull)
        let cli = try scanCLI(store, prices, cursor: &cursor, full: effectiveFull)
        markActivityCurrent(&cursor)
        try store.setScanCursor(tool: name, cursor: cursor)
        return ScanOutcome(added: journal.added + cli.added,
                           updated: journal.updated + cli.updated,
                           files: journal.files + cli.files,
                           activityAdded: journal.activityAdded + cli.activityAdded,
                           activityUpdated: journal.activityUpdated + cli.activityUpdated)
    }
}

File: windows/src/TokenTracker.Core/Scanners/KimiScanner.cs
using TokenTracker.Core.Activity;
using TokenTracker.Core.Json;
using TokenTracker.Core.Platform;
using TokenTracker.Core.Pricing;
using TokenTracker.Core.Store;

namespace TokenTracker.Core.Scanners;

/// <summary>
/// 移植 KimiScanner.swift（scanners/kimi.py）：~/.kimi-code/server/events/session_*.jsonl（事件日志）
/// + kimi-cli ~/.kimi/sessions/（通用 JSONL 兜底）。turn.step.completed 的 usage 是每步增量，直接累加。
/// </summary>
public sealed class KimiScanner(string journalDir, string cliDir) : IScanner
{
    public string Name => "kimi";
    public string Detail => "~/.kimi-code/server/events/session_*.jsonl";
    public string JournalDir { get; } = WinPaths.Expand(journalDir);
    public string CliDir { get; } = WinPaths.Expand(cliDir);

    public bool Detect() => Directory.Exists(JournalDir) || Directory.Exists(CliDir);

    /// <summary>只取公开的 Skill 标识；事件参数/内容绝不入库。</summary>
    static string SkillName(Dictionary<string, object?> payload)
    {
        var value = payload.Get("skill");
        if (value is Dictionary<string, object?> skill)
            value = skill.Get("name") ?? skill.Get("skillName") ?? skill.Get("id");
        return PyJson.Identifier(payload.Get("skillName"), payload.Get("name"), value, payload.Get("skill_name"));
    }

    static List<string> JsonlFiles(string root, Func<string, bool> predicate) =>
        WinPaths.EnumerateFilesRelative(root, predicate).Select(rel => WinPaths.JoinRelative(root, rel)).ToList();

    ScanOutcome ScanJournal(UsageStore store, PriceTable prices, Dictionary<string, object?> cursor, bool full)
    {
        var outcome = new ScanOutcome();
        if (!Directory.Exists(JournalDir)) return outcome;
        var files = JsonlFiles(JournalDir, rel =>
        {
            var name = WinPaths.LastComponent(rel);
            return name.StartsWith("session_", StringComparison.Ordinal) && rel.EndsWith(".jsonl", StringComparison.Ordinal);
        });
        foreach (var path in files)
        {
            if (!full && !FileIdentity.Changed(cursor, path)) continue;
            if (FileIdentity.Of(path) is not { } statKey) continue;
            outcome.Files++;
            var filename = WinPaths.LastComponent(path);
            var sessionId = filename["session_".Length..^".jsonl".Length];
            var project = "";
            var modelHint = "";
            string? title = null;
            foreach (var (_, obj) in ScannerSupport.IterJsonl(path))
            {
                var kind = obj.Get("kind") as string;
                var envelope = obj.GetDict("envelope") ?? new Dictionary<string, object?>();
                var payload = envelope.GetDict("payload") ?? new Dictionary<string, object?>();
                var eventType = envelope.Get("type") as string ?? "";
                var activityTs = ScannerSupport.ParseTs(PyJson.Or(envelope.Get("timestamp"), obj.Get("time")));
                long? startedAt = activityTs == 0 ? null : activityTs;
                var turnId = payload.Get("turnId") as string ?? "";
                if (kind == "event" && eventType == "skill.activated")
                {
                    var name = SkillName(payload);
                    if (name.Trim().Length > 0)
                    {
                        var callId = PyJson.Identifier(payload.Get("id"), payload.Get("skillCallId"), obj.Get("seq"));
                        var change = store.RecordActivity(Name, $"{sessionId}|skill|{callId}", "Skill", sessionId,
                            turnId, callId, startedAt: startedAt, sourceKind: "kimi_skill_activated",
                            confidence: "exact", arguments: new Dictionary<string, object?> { ["skill"] = name },
                            eventKind: ActivityKind.Skill);
                        outcome.ActivityAdded += change.Added;
                        outcome.ActivityUpdated += change.Updated;
                    }
                }
                else if (kind == "event" && eventType.StartsWith("subagent.", StringComparison.Ordinal))
                {
                    var callId = PyJson.Identifier(payload.Get("id"), payload.Get("agentId"), obj.Get("seq"));
                    var parent = PyJson.OrString(payload.Get("parentId"), payload.Get("parentAgentId"));
                    var change = store.RecordActivity(Name, $"{sessionId}|agent|{callId}",
                        PyJson.OrString(payload.Get("name"), eventType), sessionId, turnId, callId, parent,
                        startedAt: startedAt, sourceKind: "kimi_subagent", eventKind: ActivityKind.Agent,
                        eventLayer: ActivityLayer.Lifecycle);
                    outcome.ActivityAdded += change.Added;
                    outcome.ActivityUpdated += change.Updated;
                }
                else if (kind == "event" && eventType == "tool.call.started")
                {
                    var call = payload.GetDict("toolCall") ?? payload;
                    var rawName = PyJson.OrString(call.Get("name"), call.Get("toolName"), call.Get("tool"));
                    var callId = PyJson.OrString(call.Get("id"), call.Get("toolCallId"), payload.Get("toolCallId"));
                    if (rawName.Length > 0)
                    {
                        var change = store.RecordActivity(Name,
                            $"{sessionId}|tool|{(callId.Length == 0 ? PyJson.PyStr(obj.Get("seq")) : callId)}",
                            rawName, sessionId, turnId, callId, startedAt: startedAt, sourceKind: "kimi_journal",
                            arguments: PyJson.Or(call.Get("args"), call.Get("arguments"), call.Get("input")));
                        outcome.ActivityAdded += change.Added;
                        outcome.ActivityUpdated += change.Updated;
                    }
                }
                else if (kind == "event" && eventType == "tool.result")
                {
                    var callId = PyJson.OrString(payload.Get("toolCallId"), payload.Get("callId"), payload.Get("id"));
                    var status = ActivityNormalizer.Status(PyJson.Or(payload.Get("status"), payload.Get("error")));
                    if (status == "unknown") status = payload.Get("error") is null ? "success" : "error";
                    outcome.ActivityUpdated += store.CompleteActivity(Name, callId, status, startedAt,
                        PyJson.AsLong(payload.Get("durationMs")));
                }
                if (title is null && kind == "event" && eventType == "turn.started"
                    && payload.Get("prompt") is string prompt && prompt.Trim().Length > 0)
                    title = TextUtil.CleanSessionTitle(prompt);
                if (kind == "event" && eventType == "event.session.created")
                {
                    var meta = payload.GetDict("session").GetDict("metadata");
                    if (meta.Get("cwd") is string { Length: > 0 } cwd) project = cwd;
                    continue;
                }
                if (kind != "event" || eventType != "turn.step.completed" || payload.GetDict("usage") is not { } usage)
                    continue;
                var inp = PyJson.JInt(usage.Get("inputOther")); // 非缓存输入增量
                var outp = PyJson.JInt(usage.Get("output"));
                var cr = PyJson.JInt(usage.Get("inputCacheRead"));
                var cw = PyJson.JInt(usage.Get("inputCacheCreation"));
                if (inp + outp + cr + cw == 0) continue;
                var m = payload.Get("model");
                if (m is Dictionary<string, object?> md) m = md.Get("id");
                if (m is string { Length: > 0 } mStr) modelHint = mStr;
                // 事件日志不携带模型字段，默认 kimi-code（k3 家族）
                var model = modelHint.Length == 0 ? "kimi-code" : modelHint;
                var key = $"{sessionId}|step|{PyJson.PyStr(obj.Get("seq"))}";
                var cost = prices.Cost(model, inp, outp, cr, cw);
                outcome.Added += store.PutEvent(Name, key, sessionId, project, activityTs, model, inp, outp, cr, cw, cost);
            }
            cursor[path] = statKey.AsDict();
            if (title is not null) store.SetSessionTitle(Name, sessionId, title);
        }
        return outcome;
    }

    ScanOutcome ScanCli(UsageStore store, PriceTable prices, Dictionary<string, object?> cursor, bool full)
    {
        var outcome = new ScanOutcome();
        if (!Directory.Exists(CliDir)) return outcome;
        foreach (var path in JsonlFiles(CliDir, rel => rel.EndsWith(".jsonl", StringComparison.Ordinal)))
        {
            if (!full && !FileIdentity.Changed(cursor, path)) continue;
            if (FileIdentity.Of(path) is not { } statKey) continue;
            outcome.Files++;
            var filename = WinPaths.LastComponent(path);
            foreach (var (lineNo, obj) in ScannerSupport.IterJsonl(path))
            {
                if (obj.GetDict("usage") is not { } usage) continue;
                var inp = PyJson.OrInt(usage.Get("input"), usage.Get("input_tokens"));
                var outp = PyJson.OrInt(usage.Get("output"), usage.Get("output_tokens"));
                if (inp + outp == 0) continue;
                var model = obj.Get("model") as string ?? "";
                var cost = prices.Cost(model, inp, outp);
                outcome.Added += store.PutEvent(Name, $"cli|{path}|{lineNo}", filename[..^".jsonl".Length],
                    WinPaths.Parent(path), ScannerSupport.ParseTs(obj.Get("timestamp")), model, inp, outp, cost: cost);
            }
            cursor[path] = statKey.AsDict();
        }
        return outcome;
    }

    public ScanOutcome Scan(UsageStore store, PriceTable prices, bool full)
    {
        var cursor = store.GetScanCursor(Name);
        var effectiveFull = full || ActivityNormalizer.NeedsBackfill(cursor);
        var journal = ScanJournal(store, prices, cursor, effectiveFull);
        var cli = ScanCli(store, prices, cursor, effectiveFull);
        ActivityNormalizer.MarkCurrent(cursor);
        store.SetScanCursor(Name, cursor);
        return new ScanOutcome
        {
            Added = journal.Added + cli.Added,
            Updated = journal.Updated + cli.Updated,
            Files = journal.Files + cli.Files,
            ActivityAdded = journal.ActivityAdded + cli.ActivityAdded,
            ActivityUpdated = journal.ActivityUpdated + cli.ActivityUpdated,
        };
    }
}
